src/parser/smt2/smt2_lexer_new.cpp

```diff
@@ -0,0 +1,385 @@
+          if (!parseNonEmptyCharList(CharacterClass::HEXADECIMAL_DIGIT))
+          {
+            return Token::NONE;
+          }
+          return Token::HEX_LITERAL;
+        case 'f':
+          pushToToken(ch);
+          // parse [0-9]+m[0-9]+
+          if (!parseNonEmptyCharList(CharacterClass::DECIMAL_DIGIT))
+          {
+            return Token::NONE;
+          }
+          if (!parseLiteralChar('m'))
+          {
+            return Token::NONE;
+          }
+          if (!parseNonEmptyCharList(CharacterClass::DECIMAL_DIGIT))
+          {
+            return Token::NONE;
+          }
+          return Token::FIELD_LITERAL;
+        default:
+          // otherwise error
+          return Token::NONE;
+      }
+      break;
+    case '"':
+      for (;;)
+      {
+        ch = nextChar();
+        if (ch == EOF)
+        {
+          return Token::UNTERMINATED_STRING_LITERAL;
+        }
+        else if (ch == '"')
+        {
+          pushToToken(ch);
+          ch = nextChar();
+          // "" denotes the escape sequence for "
+          if (ch != '"')
+          {
+            saveChar(ch);
+            return Token::STRING_LITERAL;
+          }
+        }
+        pushToToken(ch);
+      }
+      break;
+    case ':':
+      // parse a simple symbol
+      if (!parseChar(CharacterClass::SYMBOL_START))
+      {
+        return Token::NONE;
+      }
+      parseNonEmptyCharList(CharacterClass::SYMBOL);
+      return Token::KEYWORD;
+    default:
+      if (isCharacterClass(ch, CharacterClass::DECIMAL_DIGIT))
+      {
+        Token res = Token::INTEGER_LITERAL;
+        // parse [0-9]*
+        parseCharList(CharacterClass::DECIMAL_DIGIT);
+        // maybe .[0-9]+
+        ch = nextChar();
+        if (ch == '.')
+        {
+          pushToToken(ch);
+          res = Token::DECIMAL_LITERAL;
+          // parse [0-9]+
+          if (!parseNonEmptyCharList(CharacterClass::DECIMAL_DIGIT))
+          {
+            return Token::NONE;
+          }
+        }
+        else
+        {
+          // otherwise, undo
+          saveChar(ch);
+        }
+        return res;
+      }
+      else if (isCharacterClass(ch, CharacterClass::SYMBOL_START))
+      {
+        // otherwise, we are a simple symbol or standard alphanumeric token
+        // note that we group the case when `:` is here.
+        parseCharList(CharacterClass::SYMBOL);
+        // tokenize the current symbol, which may be a special case
+        return tokenizeCurrentSymbol();
+      }
+      // otherwise error
+      break;
+  }
+  return Token::NONE;
+}
+
+char Smt2LexerNew::nextChar()
+{
+  char res;
+  if (d_peekedChar)
+  {
+    res = d_chPeeked;
+    d_peekedChar = false;
+  }
+  else
+  {
+    res = readNextChar();
+    if (res == '\n')
+    {
+      d_span.d_end.d_line++;
+      d_span.d_end.d_column = 0;
+    }
+    else
+    {
+      d_span.d_end.d_column++;
+    }
+  }
+  return res;
+}
+
+void Smt2LexerNew::saveChar(char ch)
```

Comment:
Moving this to the header may speed up lexing.

Reply:
This one is called relatively rarely.

Reply:
It did make a difference for the Bitwuzla Lexer. Have you tried or you didn't based on your conjecture that it is called rarely? (The definition of "rarely" in this context may be way less rare than you think.)

Reply:
It's called at most 1 per token I think, but regardless I went ahead and moved it since its a small function.